### PR TITLE
Xcode 8 + Swift 3

### PR DESCRIPTION
**Pull request by mohpor**

### CODE REVIEW

**Comment by mohpor** [comment]
Some notes:

I have tested the framework locally on my computer and again on my devices, ran the test and everything went fine, all tests' passed. Except, when actually testing it on device, the TwitterBot example would crash upon displaying the action sheet, complaining about internal inconsistency with number of rows and the data set.

It's worth taking a look.

File: Source/ActionController.swift [hunk 1]
```diff
@@ -265,340 +266,340 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
...
-    public func numberOfSectionsInCollectionView(collectionView: UICollectionView) -> Int {
+    open func numberOfSections(in collectionView: UICollectionView) -> Int {
         return numberOfSections()
     }
     
-    public func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+    open func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
         if hasHeader() && section == 0 { return 0 }
-        if settings.behavior.useDynamics && section > _dynamicSectionIndex {
+        if settings.behavior.useDynamics && section > _dynamicSectionIndex ?? 0 {
```
Comment (m-revetria) [inline Source/ActionController.swift]: This is not exactly the same, now if `_dynamicSectionIndex` is null and `section` is 0, then this function si returning `_sections[actionSectionIndexFor(section)].actions.count`. While previously it returned 0.

Additionally, this is causing the crash that you saw. When `_dynamicSectionIndex` is null, it means that no data has been added to the collection view (in the case `useDynamics==true`). So the collection view datasource is returning x number of rows for section 0 when no one has been inserted.

Next code should be fix the crash:
```swift
    open func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        if hasHeader() && section == 0 { return 0 }
        let rows = _sections[actionSectionIndexFor(section)].actions.count

        guard let dynamicSectionIndex = _dynamicSectionIndex else {
            return settings.behavior.useDynamics ? 0 : rows
        }
        if settings.behavior.useDynamics && section > dynamicSectionIndex {
            return 0
        }
        return rows
    }
```

File: Source/ActionController.swift [hunk 2]
```diff
@@ -265,340 +266,340 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
...
+    open func collectionView(_ collectionView: UICollectionView, shouldSelectItemAt indexPath: IndexPath) -> Bool {
         return !disableActions && actionForIndexPath(actionIndexPathFor(indexPath))?.enabled == true
     }
     
-    public func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
+    open func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
         let action = self.actionForIndexPath(actionIndexPathFor(indexPath))
 
-        if let action = action where action.executeImmediatelyOnTouch {
+        if let action = action , action.executeImmediatelyOnTouch {
```
Comment (m-revetria) [inline Source/ActionController.swift]: please remove the white space before the comma

File: Example/CustomActionControllers/Twitter.swift [hunk 3]
```diff
@@ -98,59 +98,63 @@ public class TwitterActionControllerHeader: UICollectionReusableView {
...
         }
         onConfigureCellForAction = { [weak self] cell, action, indexPath in
             cell.setup(action.data?.title, detail: action.data?.subtitle, image: action.data?.image)
-            cell.separatorView?.hidden = indexPath.item == (self?.collectionView.numberOfItemsInSection(indexPath.section))! - 1
+            cell.separatorView?.isHidden = indexPath.item == (self?.collectionView.numberOfItems(inSection: indexPath.section))! - 1
             cell.alpha = action.enabled ? 1.0 : 0.5
         }
     }
+  
+  required public init?(coder aDecoder: NSCoder) {
```
Comment (m-revetria) [inline Example/CustomActionControllers/Twitter.swift]: please indent using 4 spaces

File: Example/CustomActionControllers/Youtube.swift [hunk 4]
```diff
@@ -91,16 +91,20 @@ public class YoutubeActionController: ActionController<YoutubeCell, ActionData,
...
         
-        cellSpec = .NibFile(nibName: "YoutubeCell", bundle: NSBundle(forClass: YoutubeCell.self), height: { _  in 46 })
+        cellSpec = .nibFile(nibName: "YoutubeCell", bundle: Bundle(for: YoutubeCell.self), height: { _  in 46 })
         
         onConfigureCellForAction = { cell, action, indexPath in
             cell.setup(action.data?.title, detail: action.data?.subtitle, image: action.data?.image)
             cell.alpha = action.enabled ? 1.0 : 0.5
             
-            UIView.animateWithDuration(0.30) {
+          UIView.animate(withDuration: 0.30) {
```
Comment (m-revetria) [inline Example/CustomActionControllers/Youtube.swift]: please fix indentation

File: Source/ActionController.swift [hunk 5]
```diff
@@ -179,84 +179,85 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
...
+                collectionView.register(UINib(nibName: nibName, bundle: bundle), forSupplementaryViewOfKind:UICollectionElementKindSectionHeader, withReuseIdentifier: ReusableViewIds.SectionHeader.rawValue)
             }
         }
         
         view.addSubview(collectionView)
         
         // calculate content Inset
         collectionView.layoutSubviews()
-        if let section = _sections.last where !settings.behavior.useDynamics {
+        if let section = _sections.last , !settings.behavior.useDynamics {
```
Comment (m-revetria) [inline Source/ActionController.swift]: please delete white space before comma

File: Source/ActionController.swift [hunk 6]
```diff
@@ -265,340 +266,340 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
...
     }
 
-    override public func viewDidLayoutSubviews() {
+    override open func viewDidLayoutSubviews() {
         super.viewDidLayoutSubviews()
         collectionView.collectionViewLayout.invalidateLayout()
     }
     
-    public override func prefersStatusBarHidden() -> Bool {
+    open override var prefersStatusBarHidden : Bool {
```
Comment (m-revetria) [inline Source/ActionController.swift]: please delete white space before colon

File: Source/ActionController.swift [hunk 7]
```diff
@@ -265,340 +266,340 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
...
         return !settings.statusBar.showStatusBar
     }
     
-    public  override func preferredStatusBarStyle() -> UIStatusBarStyle {
+    open  override var preferredStatusBarStyle : UIStatusBarStyle {
```
Comment (m-revetria) [inline Source/ActionController.swift]: here too and place remove extra space between `open` and `override`

File: Source/ActionController.swift [hunk 8]
```diff
@@ -265,340 +266,340 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
...
         let action = self.actionForIndexPath(actionIndexPathFor(indexPath))
 
-        if let action = action where action.executeImmediatelyOnTouch {
+        if let action = action , action.executeImmediatelyOnTouch {
             action.handler?(action)
         }
 
         self.dismiss() {
-            if let action = action where !action.executeImmediatelyOnTouch {
+            if let action = action , !action.executeImmediatelyOnTouch {
```
Comment (m-revetria) [inline Source/ActionController.swift]: please remove space before comma

**Comment by 8masterofpuppets8** [comment]
Just noticed this PR and was wondering if it will get merged in at some point? This is the relevant issue for this #25

**Comment by ryanholden8** [comment]
Looking forward to the merge, great work!

**Comment by thonydam** [comment]
👍

File: Source/ActionController.swift (hunk 5 above)
Comment (mohpor) [inline Source/ActionController.swift]: This is caused by the Xcode migrator tool which replaced `where` with `,` (Notice that there is a whitespace before and after `where` but there shouldn't be white space there. I shouldn't have overlooked these. Sorry)

File: Example/CustomActionControllers/Youtube.swift (hunk 4 above)
Comment (mohpor) [inline Example/CustomActionControllers/Youtube.swift]: Sure, But I fail to see the point of an empty animation block! Did you mean to put that alpha value change inside the animation blog?

**Comment by mohpor** [comment]
@m-revetria, I have rebased the PR based on your comments.
The error has been fixed for TwitterBot example thanks to the code you've provided.
(Honestly, I have converted so many Libraries to Swift 3 in the past couple of weeks, my mind's playing tricks on me and makes me do stupid things like coercing nulls into 0! sorry about that)

**Comment by m-revetria** [comment]
@mohpor great work! thanks for your help and your time updating this PR!

**Pull request merged**